File: apps/api/app/schemas/alert_settings.py
from typing import Optional

from pydantic import AnyHttpUrl, BaseModel, Field


class AlertChannels(BaseModel):
    in_app: bool
    email: bool
    webhook: bool


def default_alert_channels() -> AlertChannels:
    return AlertChannels(in_app=True, email=False, webhook=False)


class AlertSettings(BaseModel):
    enabled: bool = True
    default_notify_days_before: int = Field(default=7, ge=1, le=365)
    channels: AlertChannels = Field(default_factory=default_alert_channels)
    webhook_url: Optional[AnyHttpUrl] = None


class AlertSettingsPatch(BaseModel):
    enabled: bool = True
    default_notify_days_before: int = Field(default=7, ge=1, le=365)
    channels: AlertChannels = Field(default_factory=default_alert_channels)
    webhook_url: Optional[AnyHttpUrl] = None


class AlertSettingsResponse(BaseModel):
    enabled: bool
    default_notify_days_before: int
    channels: AlertChannels
    webhook_url_preview: Optional[str] = None
    webhook_configured: bool

File: apps/web/src/lib/api/alert-settings.ts
import { request } from "./client"

export interface AlertChannels {
  in_app: boolean
  email: boolean
  webhook: boolean
}

export interface AlertSettingsResponse {
  enabled: boolean
  default_notify_days_before: number
  channels: AlertChannels
  webhook_url_preview: string | null
  webhook_configured: boolean
}

export interface AlertSettingsUpdate {
  enabled?: boolean
  default_notify_days_before?: number
  channels?: AlertChannels
  webhook_url?: string | null
}

export interface TestWebhookResponse {
  success: boolean
}

export const alertSettingsApi = {
  get(projectId: string): Promise<AlertSettingsResponse> {
    return request("GET", `/api/v1/projects/${projectId}/alert-settings`)
  },

  update(projectId: string, body: AlertSettingsUpdate): Promise<AlertSettingsResponse> {
    return request("PATCH", `/api/v1/projects/${projectId}/alert-settings`, body)
  },

  testWebhook(projectId: string): Promise<TestWebhookResponse> {
    return request("POST", `/api/v1/projects/${projectId}/alert-settings/test-webhook`)
  },
}
